core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala

```diff
@@ -311,16 +313,46 @@ class BlockManagerMasterEndpoint(
   }
 
   private def removeShuffle(shuffleId: Int): Future[Seq[Boolean]] = {
-    // Nothing to do in the BlockManagerMasterEndpoint data structures
+    // Find all shuffle blocks on executors that are no longer running
+    val blocksToDeleteByShuffleService =
```

Comment:
So, it seems here that an executor needs to be active on the node which triggers the local shuffle service to delete shuffle blocks of all the previous released executors. Is that correct? For nodes which don't have any active executors, are their shuffle files deleted only when the application stops?

Reply:
Are you asking about existing behavior or changes made here? This removes blocks directly through the shuffle service, so no executors need to be running on that node to delete the shuffle files. Obviously the shuffle service has to still be running

Reply:
Ok. I misunderstood. This code is run by the driver.

Reply:
Ah yep